File: validators/javascript/validate.js
#!/usr/bin/env node
/**
 * Aifeels Conformance Test Suite - JavaScript Validator
 * Validates JavaScript/TypeScript implementations against official test vectors.
 */

const fs = require('fs');
const path = require('path');

class ConformanceValidator {
  /**
   * Initialize validator.
   * @param {string} testVectorsPath - Path to test-vectors.json
   * @param {string} implementationPath - Path to implementation module
   */
  constructor(testVectorsPath, implementationPath) {
    this.testData = JSON.parse(fs.readFileSync(testVectorsPath, 'utf8'));
    this.implementation = require(implementationPath);
    this.results = [];
  }

  /**
   * Get nested value using dot notation (e.g., 'metadata.trend').
   * @param {object} obj - Object to traverse
   * @param {string} path - Dot-separated path
   * @returns {*} Value at path
   */
  getNestedValue(obj, path) {
    const keys = path.split('.');
    let value = obj;
    for (const key of keys) {
      if (value && typeof value === 'object' && key in value) {
        value = value[key];
      } else {
        throw new Error(`Cannot access '${key}' in object`);
      }
    }
    return value;
  }

  /**
   * Check if assertion passes.
   * @param {object} state - State object to check
   * @param {object} assertion - Assertion definition
   * @returns {boolean} True if assertion passes
   */
  assertValue(state, assertion) {
    const { path, expected, type, tolerance = 0.001 } = assertion;

    let actual;
    try {
      actual = this.getNestedValue(state, path);
    } catch (e) {
      console.log(`  ✗ Cannot access ${path}: ${e.message}`);
      return false;
    }

    if (type === 'equals') {
      if (actual !== expected) {
        console.log(`  ✗ ${path}: expected ${expected}, got ${actual}`);
        return false;
      }
    } else if (type === 'approximately') {
      if (Math.abs(actual - expected) > tolerance) {
        console.log(`  ✗ ${path}: expected ~${expected}, got ${actual} (tolerance: ${tolerance})`);
        return false;
      }
    } else {
      console.log(`  ✗ Unknown assertion type: ${type}`);
      return false;
    }

    return true;
  }

  /**
   * Run a single conformance test.
   * @param {object} test - Test definition
   * @returns {boolean} True if test passes
   */
  runTest(test) {
    const { id, name, setup, steps, assertions } = test;

    console.log(`\nRunning ${id}: ${name}...`);

    try {
      // Setup
      let state;
      if (setup.action === 'initialize') {
        state = new this.implementation.EmotionalState();
        if (setup.initial_state) {
          // Set initial state values
          for (const [key, value] of Object.entries(setup.initial_state)) {
            state[key] = value;
          }
        }
      } else {
        console.log(`  ✗ Unknown setup action: ${setup.action}`);
        return false;
      }

      // Execute steps
      for (const step of steps) {
        const { action } = step;

        if (action === 'process_event') {
          state.processEvent(step.event);
        } else if (action === 'advance_time') {
          const seconds = step.seconds;
          // Try advance_time, otherwise fall back to apply_decay
          if (typeof state.advanceTime === 'function') {
            state.advanceTime(seconds);
          } else {
            // Calculate intervals and apply decay
            const intervals = Math.floor(seconds / 300);
            for (let i = 0; i < intervals; i++) {
              state.applyDecay();
            }
          }
        } else if (action === 'get_recommended_action') {
          // Store for assertion checking
          state._recommendedAction = state.recommendedAction();
        } else {
          console.log(`  ✗ Unknown step action: ${action}`);
          return false;
        }
      }

      // Check assertions
      let allPassed = true;
      for (const assertion of assertions) {
        // Handle recommended_action specially
        if (assertion.path === 'recommended_action') {
          if (!('_recommendedAction' in state)) {
            console.log(`  ✗ recommended_action not set`);
            allPassed = false;
            continue;
          }
          const actual = state._recommendedAction;
          const expected = assertion.expected;
          if (actual !== expected) {
            console.log(`  ✗ recommended_action: expected ${expected}, got ${actual}`);
            allPassed = false;
          }
          continue;
        }

        if (!this.assertValue(state, assertion)) {
          allPassed = false;
        }
      }

      if (allPassed) {
        console.log(`✓ ${id}: ${name} - PASSED`);
        return true;
      } else {
        console.log(`✗ ${id}: ${name} - FAILED`);
        return false;
      }

    } catch (e) {
      console.log(`✗ ${id}: ${name} - ERROR: ${e.message}`);
      console.error(e.stack);
      return false;
    }
  }

  /**
   * Run all conformance tests.
   * @returns {boolean} True if all tests pass
   */
  runAllTests() {
    console.log(`Aifeels Conformance Test Suite v${this.testData.version}`);
    console.log(`Spec version: ${this.testData.spec_version}`);
    console.log(`Testing implementation: ${this.implementation.name || 'unknown'}`);
    console.log('='.repeat(60));

    let passed = 0;
    let failed = 0;

    for (const test of this.testData.tests) {
      const result = this.runTest(test);
      this.results.push({
        id: test.id,
        name: test.name,
        status: result ? 'PASSED' : 'FAILED'
      });

      if (result) {
        passed++;
      } else {
        failed++;
      }
    }

    console.log('\n' + '='.repeat(60));
    console.log(`Results: ${passed} passed, ${failed} failed out of ${passed + failed} total`);

    if (failed === 0) {
      console.log('✓ CONFORMANT: Implementation passes all tests!');
      return true;
    } else {
      console.log('✗ NON-CONFORMANT: Implementation failed one or more tests.');
      return false;
    }
  }

  /**
   * Generate conformance report JSON.
   * @param {string} outputPath - Path to write report
   */
  generateReport(outputPath) {
    const total = this.results.length;
    const passed = this.results.filter(r => r.status === 'PASSED').length;
    const failed = total - passed;

    const report = {
      implementation: {
        name: this.implementation.name || 'unknown',
        version: this.implementation.version || 'unknown',
        language: 'JavaScript',
        license: 'Apache-2.0'
      },
      spec_version: this.testData.spec_version,
      test_suite_version: this.testData.version,
      test_results: {
        total,
        passed,
        failed,
        errors: 0,
        pass_rate: total > 0 ? (passed / total * 100) : 0
      },
      test_details: this.results,
      conformance_statement: (
        `This implementation is ${failed === 0 ? 'fully conformant' : 'NOT conformant'} ` +
        `with the Aifeels Specification v${this.testData.spec_version}.`
      )
    };

    fs.writeFileSync(outputPath, JSON.stringify(report, null, 2));
    console.log(`\nReport written to: ${outputPath}`);
  }
}

function main() {
  if (process.argv.length < 3) {
    console.log('Usage: node validate.js <implementation_path>');
    console.log('Example: node validate.js ./aifeels.js');
    process.exit(1);
  }

  const implementationPath = path.resolve(process.argv[2]);

  // Find test vectors
  const scriptDir = __dirname;
  const testVectorsPath = path.join(scriptDir, '..', '..', 'test-vectors', 'test-vectors.json');

  if (!fs.existsSync(testVectorsPath)) {
    console.error(`Error: test-vectors.json not found at ${testVectorsPath}`);
    process.exit(1);
  }

  // Run validation
  const validator = new ConformanceValidator(testVectorsPath, implementationPath);
  const success = validator.runAllTests();

  // Generate report
  const reportPath = path.join(scriptDir, 'conformance-report.json');
  validator.generateReport(reportPath);

  process.exit(success ? 0 : 1);
}

if (require.main === module) {
  main();
}

module.exports = { ConformanceValidator };

File: validators/python/validate.py
#!/usr/bin/env python3
"""
Aifeels Conformance Test Suite - Python Validator
Validates Python implementations against official test vectors.
"""

import json
import sys
import importlib
import os
from pathlib import Path
from typing import Any, Dict, List, Optional


class ConformanceValidator:
    """Validates Aifeels implementations against test vectors."""
    
    def __init__(self, test_vectors_path: str, implementation_module: str):
        """
        Initialize validator.
        
        Args:
            test_vectors_path: Path to test-vectors.json
            implementation_module: Python module name to test
        """
        with open(test_vectors_path, 'r') as f:
            self.test_data = json.load(f)
        
        self.implementation = importlib.import_module(implementation_module)
        self.results = []
        
    def get_nested_value(self, obj: Any, path: str) -> Any:
        """Get nested value using dot notation (e.g., 'metadata.trend')."""
        keys = path.split('.')
        value = obj
        for key in keys:
            if hasattr(value, key):
                value = getattr(value, key)
            elif isinstance(value, dict):
                value = value[key]
            else:
                raise AttributeError(f"Cannot access '{key}' in {type(value)}")
        return value
    
    def assert_value(self, state: Any, assertion: Dict) -> bool:
        """Check if assertion passes."""
        path = assertion['path']
        expected = assertion['expected']
        assert_type = assertion['type']
        
        try:
            actual = self.get_nested_value(state, path)
        except (AttributeError, KeyError) as e:
            print(f"  ✗ Cannot access {path}: {e}")
            return False
        
        if assert_type == 'equals':
            if actual != expected:
                print(f"  ✗ {path}: expected {expected}, got {actual}")
                return False
        elif assert_type == 'approximately':
            tolerance = assertion.get('tolerance', 0.001)
            if abs(actual - expected) > tolerance:
                print(f"  ✗ {path}: expected ~{expected}, got {actual} (tolerance: {tolerance})")
                return False
        else:
            print(f"  ✗ Unknown assertion type: {assert_type}")
            return False
        
        return True
    
    def run_test(self, test: Dict) -> bool:
        """Run a single conformance test."""
        test_id = test['id']
        test_name = test['name']
        
        print(f"\nRunning {test_id}: {test_name}...")
        
        try:
            # Setup
            setup = test['setup']
            if setup['action'] == 'initialize':
                state = self.implementation.EmotionalState()
                if 'initial_state' in setup:
                    # Set initial state values
                    init_state = setup['initial_state']
                    for key, value in init_state.items():
                        setattr(state, key, value)
            else:
                print(f"  ✗ Unknown setup action: {setup['action']}")
                return False
            
            # Execute steps
            for step in test['steps']:
                action = step['action']
                
                if action == 'process_event':
                    event = step['event']
                    state.process_event(event)
                elif action == 'advance_time':
                    seconds = step['seconds']
                    # Assuming implementation has advance_time or similar
                    if hasattr(state, 'advance_time'):
                        state.advance_time(seconds)
                    else:
                        # Calculate intervals and apply decay
                        intervals = seconds // 300
                        for _ in range(intervals):
                            state.apply_decay()
                elif action == 'get_recommended_action':
                    # Store for assertion checking
                    state._recommended_action = state.recommended_action()
                else:
                    print(f"  ✗ Unknown step action: {action}")
                    return False
            
            # Check assertions
            all_passed = True
            for assertion in test['assertions']:
                # Handle recommended_action specially
                if assertion['path'] == 'recommended_action':
                    if not hasattr(state, '_recommended_action'):
                        print(f"  ✗ recommended_action not set")
                        all_passed = False
                        continue
                    actual = state._recommended_action
                    expected = assertion['expected']
                    if actual != expected:
                        print(f"  ✗ recommended_action: expected {expected}, got {actual}")
                        all_passed = False
                    continue
                
                if not self.assert_value(state, assertion):
                    all_passed = False
            
            if all_passed:
                print(f"✓ {test_id}: {test_name} - PASSED")
                return True
            else:
                print(f"✗ {test_id}: {test_name} - FAILED")
                return False
                
        except Exception as e:
            print(f"✗ {test_id}: {test_name} - ERROR: {e}")
            import traceback
            traceback.print_exc()
            return False
    
    def run_all_tests(self) -> bool:
        """Run all conformance tests."""
        print(f"Aifeels Conformance Test Suite v{self.test_data['version']}")
        print(f"Spec version: {self.test_data['spec_version']}")
        print(f"Testing implementation: {self.implementation.__name__}")
        print("=" * 60)
        
        passed = 0
        failed = 0
        
        for test in self.test_data['tests']:
            result = self.run_test(test)
            self.results.append({
                'id': test['id'],
                'name': test['name'],
                'status': 'PASSED' if result else 'FAILED'
            })
            
            if result:
                passed += 1
            else:
                failed += 1
        
        print("\n" + "=" * 60)
        print(f"Results: {passed} passed, {failed} failed out of {passed + failed} total")
        
        if failed == 0:
            print("✓ CONFORMANT: Implementation passes all tests!")
            return True
        else:
            print("✗ NON-CONFORMANT: Implementation failed one or more tests.")
            return False
    
    def generate_report(self, output_path: str):
        """Generate conformance report JSON."""
        total = len(self.results)
        passed = sum(1 for r in self.results if r['status'] == 'PASSED')
        failed = total - passed
        
        report = {
            "implementation": {
                "name": self.implementation.__name__,
                "version": getattr(self.implementation, '__version__', 'unknown'),
                "language": "Python",
                "license": "Apache-2.0"
            },
            "spec_version": self.test_data['spec_version'],
            "test_suite_version": self.test_data['version'],
            "test_results": {
                "total": total,
                "passed": passed,
                "failed": failed,
                "errors": 0,
                "pass_rate": (passed / total * 100) if total > 0 else 0
            },
            "test_details": self.results,
            "conformance_statement": (
                f"This implementation is {'fully conformant' if failed == 0 else 'NOT conformant'} "
                f"with the Aifeels Specification v{self.test_data['spec_version']}."
            )
        }
        
        with open(output_path, 'w') as f:
            json.dump(report, f, indent=2)
        
        print(f"\nReport written to: {output_path}")


def main():
    """Main entry point."""
    if len(sys.argv) < 2:
        print("Usage: python validate.py <implementation_module>")
        print("Example: python validate.py aifeels")
        sys.exit(1)
    
    implementation_module = sys.argv[1]
    
    # Find test vectors
    script_dir = Path(__file__).parent
    test_vectors_path = script_dir.parent.parent / 'test-vectors' / 'test-vectors.json'
    
    if not test_vectors_path.exists():
        print(f"Error: test-vectors.json not found at {test_vectors_path}")
        sys.exit(1)
    
    # Run validation
    validator = ConformanceValidator(str(test_vectors_path), implementation_module)
    success = validator.run_all_tests()
    
    # Generate report
    report_path = script_dir / 'conformance-report.json'
    validator.generate_report(str(report_path))
    
    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()
